src/proj7BittingCerratoCohenEllmer/bantam/lexer/Scanner.java

```diff
@@ -114,496 +68,77 @@ public Scanner(Reader reader, ErrorHandler handler) {
-        if (Character.isDigit(spellingStack.peek())) {
-            return false;
-        } else {
-            skippedLastToken = spellingStack.pop();
-            return true;
-        }
-    }
-
-    /**
-     * checks if string contains valid characters
-     * <p>
-     * String constants start and end with double quotes.
-     *
-     * @param spellingStack the stack containing the characters
-     * @return returns true if the stack contains a complete string
-     */
-    private boolean isCompleteString(Stack<Character> spellingStack) {
-        int stackSize = spellingStack.size();
-        if (stackSize > 1) {
-            // check if string is closed
-            if (spellingStack.peek() == '"' && !charIsEscaped) {
-                if (sourceFile.getCurrentLineNumber() != stringStart) {
-                    errorHandler.register(Error.Kind.LEX_ERROR,
-                            sourceFile.getFilename(),
-                            sourceFile.getCurrentLineNumber(),
-                            "Multiline String found! Starting @ line: "
-                                    + stringStart);
-                }
-                skippedLastToken = '\0';
-                return true;
-            }
-            //check if EOF triggers untermintated string
-            if (spellingStack.peek() == '\u0000'){
-                errorHandler.register(Error.Kind.LEX_ERROR,
-                                sourceFile.getFilename(), sourceFile.getCurrentLineNumber(),
-                                "Unterminated String Constant!");
-                return true;
-            }
-            // logic for handling escape sequences
-            if (spellingStack.peek() == '\\'){
-                charIsEscaped = true;
-                return false;
+                precursorToken.pushChar(getNextChar(false));
             }
-            if (charIsEscaped && !validEscapedCharacter.contains(spellingStack.peek())){
-                errorHandler.register(Error.Kind.LEX_ERROR,
-                                sourceFile.getFilename(), sourceFile.getCurrentLineNumber(),
-                                "Invalid Escaped Character \\" 
-                                + spellingStack.peek() + "!");
-                charIsEscaped = false;
-                return false;
-            } else{
-                charIsEscaped = false;
-                return false;
+        } while (!precursorToken.isComplete());
+
+        // store last char in skippedLastToken if needed
+        Optional<Character> extraChar = precursorToken.getExtraChar();
+        extraChar.ifPresent(c -> skippedLastToken = c);
+
+        // create the token from the precursor token
+        // create token before extracting errors as some errors occur during creation
+        Token finalToken;
+        try {
+            finalToken = precursorToken.getFinalToken(sourceFile.getCurrentLineNumber());
+        } catch (MalformedSpellingStackException e) {
```

Comment:
Im pretty confused, but trusting, of this "MalformedSpellingStackException" It seems like a bunch of repetition:

theres an "if(poplasttoken) {throw malformed}" structure i think 9 times. Which seems weird when  here we catch something that internal documentation stat "will never happen".